ENIGMAsystem/SHELL/Graphics_Systems/OpenGL1/GLvertex.cpp

```diff
@@ -15,86 +15,155 @@
+void vertex_color(int buffer, int color, double alpha) {
+  enigma::color_t finalcol = color + (bind_alpha(alpha) << 24);
+  enigma::vertexBuffers[buffer]->vertices.push_back(finalcol);
+}
 
-  void vertex_format_add_normal() {}
+void vertex_submit(int buffer, int primitive, unsigned vertex_start, unsigned vertex_count) {
+  enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[buffer];
+  const enigma::VertexFormat* vertexFormat = enigma::vertexFormats[vertexBuffer->format];
+
+  // if the contents of the vertex buffer are dirty then we need to update
+  // our native vertex buffer object "peer"
+  if (vertexBuffer->dirty) {
+    GLuint vertexBufferPeer;
+    auto it = enigma::vertexBufferPeers.find(buffer);
+    size_t size = enigma_user::vertex_get_size(buffer);
+
+    // if we haven't created a native "peer" vbo for this vertex buffer yet,
+    // then we need to do so now
+    if (it == enigma::vertexBufferPeers.end()) {
+      glGenBuffers(1, &vertexBufferPeer);
+      enigma::vertexBufferPeers[buffer] = vertexBufferPeer;
+    } else {
+      vertexBufferPeer = it->second;
+    }
+
+    glBindBuffer(GL_ARRAY_BUFFER, vertexBufferPeer);
+    GLint pSize;
+    glGetBufferParameteriv(GL_ARRAY_BUFFER, GL_BUFFER_SIZE, &pSize);
+
+    // if the size of the peer vbo isn't big enough to hold the new contents
+    // or vertex_freeze was called, then we need to make a call to glBufferData
+    // to allocate a bigger peer vbo or remove the GL_DYNAMIC_DRAW usage
+    if (size > pSize || vertexBuffer->frozen) {
+      GLenum usage = vertexBuffer->frozen ? GL_STATIC_DRAW : GL_DYNAMIC_DRAW;
+      glBufferData(GL_ARRAY_BUFFER, size, &vertexBuffer->vertices[0], usage);
+    } else {
+      glBufferSubData(GL_ARRAY_BUFFER, 0, size, &vertexBuffer->vertices[0]);
+    }
+
+    vertexBuffer->vertices.clear();
+    vertexBuffer->dirty = false;
+  } else {
+    glBindBuffer(GL_ARRAY_BUFFER, enigma::vertexBufferPeers[buffer]);
+  }
 
-  void vertex_format_add_custom(int type, int usage) {}
+  bool useVertices = false, useNormals = false, useColors = false, useTextCoords = false, useFogCoords = false;
+  size_t offset = 0, texture = 0;
+  const size_t stride = vertexFormat->stride * sizeof(float);
+  for (size_t i = 0; i < vertexFormat->flags.size(); ++i) {
+    const pair<int, int> flag = vertexFormat->flags[i];
+
+    size_t elements = 0, size = 0;
+    GLenum type = GL_FLOAT;
+    switch (flag.first) {
+      case vertex_type_float1: elements = 1; size = 1; break;
+      case vertex_type_float2: elements = 2; size = 2; break;
+      case vertex_type_float3: elements = 3; size = 3; break;
+      case vertex_type_float4: elements = 4; size = 4; break;
+      case vertex_type_color: elements = 4; size = 1; type = GL_UNSIGNED_BYTE; break;
+      case vertex_type_ubyte4: elements = 4; size = 1; type = GL_UNSIGNED_BYTE; break;
+    }
+
+    #define OFFSET( P )  ( ( const GLvoid * ) ( sizeof( gs_scalar ) * ( P         ) ) )
```

Comment:
what is this turd?

Reply:
When you have the format position followed by color in your vertices array, you set it up with OpenGL by specifying the offset at which colors/normals/texture coordinates start. When you use a vertex array it's treated as the literal pointer from which the attributes/vertex data is read. However, we are actually using a VBO now, so it's actually a byte offset into `GL_ARRAY_BUFFER`.

See legacy glVertexPointer documentation:
https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glVertexPointer.xml
https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glColorPointer.xml

> If a non-zero named buffer object is bound to the GL_ARRAY_BUFFER target 
            (see glBindBuffer) while a vertex array is
            specified, pointer is treated as a byte offset into the buffer object's data store.
            Also, the buffer object binding (GL_ARRAY_BUFFER_BINDING) is saved as vertex array
            client-side state (GL_VERTEX_ARRAY_BUFFER_BINDING).

Reply:
I more so meant fix format and or change it to function and write what it does "OFFSET" is very vague

Reply:
Is there a good reason to format it like ass? 😛

Reply:
Ok, I fixed both things you two were complaining about: 9d00826fc203288aba8d94a59f78fbb69f9c13f9

I renamed it to `GL_ATTRIB_OFFSET` because that better describes what it does. I don't know how it got textually formatted with a bunch of whitespace, but I fixed that. If you mean why the vertex data is formatted like shit, that's a matter of perception, interleaved vertex data is more cache friendly and as a consequent more efficient and thus why we continue to do that just like GM.